File: lab10-iterative_linear/lab10-iterative_linear_cxx/linresid.cpp
/* Daniel R. Reynolds
   SMU Mathematics
   Math 4370 / 6370 */

// Inclusions
#include <stdlib.h>
#include <cmath>


// Description: calculates the linear residual and its averaged 2-norm (WRMS)
int linresid(double *a, double *b, double *c, double *u,
	     double *r, double *res, double &norm2, int N) {

  // compute linear residual at left of subdomain
  res[0] = b[0]*u[0] + c[0]*u[1] - r[0];
  norm2 = res[0]*res[0];

  // compute linear residual in interior of subdomain
  int k;
  for (k=1; k<N-1; k++) {
    res[k] = a[k]*u[k-1] + b[k]*u[k] + c[k]*u[k+1] - r[k];
    norm2 += res[k]*res[k];
  }

  // compute linear residual at right end of subdomain
  k = N-1;
  res[k] = a[k]*u[k-1] + b[k]*u[k] - r[k];
  norm2 += res[k]*res[k];

  // compute averaged 2-norm
  norm2 = sqrt(norm2)/N;

  return 0;
} // end linresid

File: lab10-iterative_linear/lab10-iterative_linear_f90/linresid.f90
! -*- Mode: Fortran90; -*-
!-----------------------------------------------------------------
! Daniel R. Reynolds
! SMU Mathematics
! Math 4370 / 6370
!=================================================================

subroutine linresid(a, b, c, u, r, res, norm2, N, ierr)
  !-----------------------------------------------------------------
  ! Description:
  !    calculates the linear residual and its averaged 2-norm (WRMS)
  !-----------------------------------------------------------------
  !======= Inclusions ===========

  !======= Declarations =========
  implicit none

  integer, intent(in) :: N
  integer, intent(out) :: ierr
  double precision, dimension(N) :: a, b, c, u, r, res
  double precision :: norm2
  integer :: k

  !======= Internals ============

  ! initialze output to success
  ierr = 0

  ! compute linear residual at left of local subdomain
  res(1) = b(1)*u(1) + c(1)*u(2) - r(1)
  norm2 = res(1)**2

  ! compute linear residual in interior of subdomain
  do k=2,N-1
    res(k) = a(k)*u(k-1) + b(k)*u(k) + c(k)*u(k+1) - r(k)
    norm2 = norm2 + res(k)**2
  end do

  ! compute linear residual at right end of local subdomain
  res(N) = a(N)*u(N-1) + b(N)*u(N) - r(N)
  norm2 = norm2 + res(N)**2

  ! compute averaged 2-norm
  norm2 = sqrt(norm2)/N

  return

end subroutine linresid
!=================================================================
